contracts/token/ERC20WithPermit.sol

```diff
@@ -1,7 +1,232 @@
 // SPDX-License-Identifier: MIT
 
-pragma solidity 0.8.4;
+pragma solidity ^0.8.5;
 
-contract Erc20WithPermit {
-    // TODO will be implemented separately
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+import "./IERC20WithPermit.sol";
+import "./IReceiveApproval.sol";
+
+/// @title  ERC20WithPermit
+/// @notice Burnable ERC20 token with EIP2612 permit functionality. User can
+///         authorize a transfer of their token with a signature conforming
+///         EIP712 standard instead of an on-chain transaction from their
+///         address. Anyone can submit this signature on the user's behalf by
+///         calling the permit function, as specified in EIP2612 standard,
+///         paying gas fees, and possibly performing other actions in the same
+///         transaction.
+contract ERC20WithPermit is IERC20WithPermit, Ownable {
+    mapping(address => uint256) public override balanceOf;
+    mapping(address => mapping(address => uint256)) public override allowance;
+
+    mapping(address => uint256) public override nonces;
+
+    uint256 public immutable cachedChainId;
+    bytes32 public immutable cachedDomainSeparator;
+
+    // keccak256("Permit(address owner,address spender,uint256 value,uint256 nonce,uint256 deadline)");
+    bytes32 public constant override PERMIT_TYPEHASH =
+        0x6e71edae12b1b97f4d1f60370fef10105fa2faae0126114a169c64845d6126c9;
+
+    uint256 public override totalSupply;
+
+    string public override name;
+    string public override symbol;
+
+    uint8 public constant override decimals = 18;
+
+    constructor(string memory _name, string memory _symbol) {
+        name = _name;
+        symbol = _symbol;
+
+        cachedChainId = chainId();
+        cachedDomainSeparator = buildDomainSeparator();
+    }
+
+    function transfer(address recipient, uint256 amount)
+        external
+        override
+        returns (bool)
+    {
+        _transfer(msg.sender, recipient, amount);
+        return true;
+    }
+
+    function transferFrom(
+        address sender,
+        address recipient,
+        uint256 amount
+    ) external override returns (bool) {
+        uint256 currentAllowance = allowance[sender][msg.sender];
+        if (currentAllowance != type(uint256).max) {
+            require(
+                currentAllowance >= amount,
+                "Transfer amount exceeds allowance"
+            );
+            _approve(sender, msg.sender, currentAllowance - amount);
+        }
```

Comment:
Shouldn't we cover the same path in the `burnFrom` function? It also decreases allowance, so if `burnFrom` is called the allowance will drop below maxuint256, and will affect the expected functionality for transfers.

Reply:
https://github.com/thesis/solidity-contracts/pull/3/commits/ee42c58081a345f759f8f89b7022c0f70109f14d